### ISSUE TITLE
Significant degradation in message rates observed on Master. 

### ISSUE THREAD

**Issue by jladd-mlnx** [open-mpi/ompi#1831]
Opening this issue for tracking purposes. Measured with Master nightly build against 1.10.3. Possible fix on master. 

@hjelmn or @bosilca please comment. 

```
PML - Yalla
OMPI – 1.10.3
$mpirun -np 2 --map-by node --bind-to core -mca pml yalla -x MXM_RDMA_PORTS=mlx5_0:1  -mca btl_openib_if_include mlx5_0:1 /hpc/local/benchmarks/hpc-stack-gcc/install/ompi-v1.10/tests/osu-micro-benchmarks-5.2/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.2
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       4.01        4005006.11
2                       8.24        4121056.15
4                      16.39        4097311.09
8                      32.45        4055766.73
16                     64.16        4010025.24
32                    127.13        3972687.66
64                    237.04        3703703.70
128                   455.11        3555555.62
256                   860.96        3363110.99
512                  1592.23        3109815.42
1024                 2811.50        2745602.68
2048                 4972.38        2427921.16
4096                 5430.79        1325875.29
8192                 5933.54         724309.64
16384                6155.42         375697.10
32768                6328.16         193120.10
65536                6398.15          97627.95
131072               6433.23          49081.64
262144               5161.27          19688.67
524288               5731.10          10931.20
1048576              6046.06           5765.97
2097152              6215.12           2963.60
4194304              6306.30           1503.54


---------------------
PML – Yalla
OMPI – Master
$mpirun -np 2 --map-by node --bind-to core -mca pml yalla -x MXM_RDMA_PORTS=mlx5_0:1  -mca btl_openib_if_include mlx5_0:1 /hpc/local/benchmarks/hpc-stack-gcc/install/ompi-master/tests/osu-micro-benchmarks-5.2/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.2
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       1.89        1887305.40
2                       3.80        1898890.08
4                       7.56        1889678.24
8                      15.31        1914346.13
16                     30.41        1900517.95
32                     60.30        1884510.12
64                    119.99        1874796.93
128                   227.47        1777098.03
256                   454.66        1776025.43
512                   870.71        1700598.93
1024                 1599.39        1561900.54
2048                 3228.97        1576645.16
4096                 4453.33        1087237.56
8192                 5822.02         710695.44
16384                6213.84         379262.51
32768                6336.49         193374.24
65536                6403.37          97707.72
131072               6438.18          49119.44
262144               5126.38          19555.59
524288               5708.48          10888.06
1048576              6033.43           5753.92
2097152              6208.48           2960.43
4194304              6303.32           1502.83


-----------------------------
PML - OB1
OMPI – 1.10.3
$mpirun -np 2 --map-by node --bind-to core -mca pml ob1 -x MXM_RDMA_PORTS=mlx5_0:1  -mca btl_openib_if_include mlx5_0:1 /hpc/local/benchmarks/hpc-stack-gcc/install/ompi-v1.10/tests/osu-micro-benchmarks-5.2/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.2
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       3.20        3204807.28
2                       6.91        3453858.56
4                      13.82        3453858.78
8                      27.41        3426124.16
16                     54.12        3382663.90
32                    105.73        3304078.53
64                    208.55        3258655.72
128                   402.16        3141875.26
256                   780.19        3047618.99
512                  1324.49        2586903.70
1024                 2392.70        2336619.29
2048                 4147.85        2025316.48
4096                 5411.73        1321222.14
8192                 5900.16         720234.07
16384                6083.99         371337.40
32768                6329.11         193149.24
65536                6427.56          98076.78
131072               6478.69          49428.48
262144               6503.55          24809.09
524288               6517.20          12430.56
1048576              6523.66           6221.44
2097152              6526.58           3112.11
4194304              6528.46           1556.51
----------------------------------
PML – OB1
OMPI - Master
$mpirun -np 2 --map-by node -mca pml ob1 -mca btl_openib_if_include mlx5_0:1 /hpc/local/benchmarks/hpc-stack-gcc/install/ompi-master/tests/osu-micro-benchmarks-5.2/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.2
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       1.64        1636174.22
2                       4.79        2392507.23
4                       9.69        2423259.37
8                      19.08        2384926.46
16                     38.57        2410744.90
32                     75.80        2368681.59
64                    149.17        2330745.92
128                   281.28        2197461.83
256                   539.24        2106415.38
512                  1065.10        2080264.37
1024                 1807.65        1765284.91
2048                 3429.21        1674421.30
4096                 5233.04        1277597.35
8192                 5634.88         687851.71
16384                5303.44         323696.21
32768                6091.79         185906.65
65536                6392.29          97538.57
131072               6459.20          49279.78
262144               6494.59          24774.90
524288               6512.32          12421.26
1048576              6521.32           6219.21
2097152              6525.70           3111.70
4194304              6526.97           1556.15
```

**Comment by hjelmn** [comment]
@jladd-mlnx You need to confirm this affects 2.0.0 before setting a blocker there.

**Comment by hjelmn** [comment]
One of the biggest differences between 2.0.0 and master is master always has MPI_THREAD_MULTIPLE support. This will impact performance and should be fixed by the next master branch.

**Comment by bosilca** [comment]
We also need to confirm that the performance drop was not already visible before the request rework made it in.

**Comment by jladd-mlnx** [comment]
Yalla looks better on 2.x, but still degraded by ~5%. OpenIB still has performance issues, and a bug that prevents the test from completing.


PML - Yalla
OMPI - v2.x 
```
$mpirun -np 2 --map-by node --bind-to core -mca pml yalla -x MXM_RDMA_PORTS=mlx5_0:1  -mca btl_openib_if_include mlx5_0:1 /hpc/mtr_scrap/users/joshual/ompi-release/osu-micro-benchmarks-5.3/install/libexec/osu-micro-benchmarks/mpi/pt2pt/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.3
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       3.78        3781191.93
2                       7.58        3788042.56
4                      15.29        3822050.11
8                      30.19        3773720.66
16                     60.38        3773580.52
32                    120.21        3756457.15
64                    235.28        3676253.05
128                   402.59        3145267.57
256                   822.60        3213293.41
512                  1535.95        2999911.89
1024                 2714.29        2650670.98
2048                 4405.85        2151295.22
4096                 5152.63        1257966.21
8192                 5725.70         698938.48
16384                6106.27         372697.15
32768                6250.77         190758.24
65536                6332.56          96627.17
131072               6403.43          48854.28
262144               6402.59          24423.95
524288               6442.80          12288.66
1048576              6460.99           6161.68
2097152              6470.35           3085.30
4194304              6475.19           1543.81
```

```
joshual@hpchead /hpc/mtr_scrap/users/joshual/ompi-release/osu-micro-benchmarks-5.3/install/libexec/osu-micro-benchmarks/mpi/pt2pt 
$mpirun -np 2 --map-by node --bind-to core -mca pml ob1 -x MXM_RDMA_PORTS=mlx5_0:1  -mca btl_openib_if_include mlx5_0:1 /hpc/mtr_scrap/users/joshual/ompi-release/osu-micro-benchmarks-5.3/install/libexec/osu-micro-benchmarks/mpi/pt2pt/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.3
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       2.04        2042588.59
2                       4.23        2115237.51
4                       8.40        2100731.38
8                      16.82        2101930.46
16                     33.66        2103450.30
32                     66.01        2062909.05
64                    129.70        2026597.13
128                   247.53        1933860.78
256                   473.00        1847674.81
512                   867.73        1694776.11
1024                 1523.01        1487313.31
2048                 2571.69        1255709.54
4096                 3762.62         918608.56
8192                 5847.15         713762.90
[vegas36][[3628,1],1][btl_tcp_endpoint.c:800:mca_btl_tcp_endpoint_complete_connect] connect() to 21.151.70.35 failed: No route to host (113)
```

**Comment by hjelmn** [comment]
Looks like the issue is the tcp btl is being used. Add ```-mca btl self,vader,openib```

**Comment by jladd-mlnx** [comment]
@hjelmn I see that. That's odd. Violates the law of least surprise. Still some degradation. 

```
$mpirun -np 2 --map-by node --bind-to core -mca pml ob1 -mca btl openib,self -x MXM_RDMA_PORTS=mlx5_0:1  -mca btl_openib_if_include mlx5_0:1 /hpc/mtr_scrap/users/joshual/ompi-release/osu-micro-benchmarks-5.3/install/libexec/osu-micro-benchmarks/mpi/pt2pt/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.3
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       3.11        3106859.54
2                       6.37        3186949.58
4                      12.81        3203248.06
8                      25.58        3197116.28
16                     50.15        3134119.36
32                     98.67        3083294.13
64                    195.66        3057110.52
128                   377.00        2945293.94
256                   722.70        2823063.32
512                  1253.78        2448783.17
1024                 2273.13        2219848.99
2048                 3797.79        1854390.37
4096                 5168.69        1261887.62
8192                 5896.71         719813.82
16384                6024.78         367723.32
32768                6278.04         191590.67
65536                6406.25          97751.62
131072               6464.88          49323.11
262144               6498.35          24789.24
524288               6514.36          12425.15
1048576              6522.33           6220.18
2097152              6526.24           3111.95
4194304              6528.29           1556.47
```

**Comment by hjelmn** [comment]
Yeah, very odd that the tcp btl is active. Is this system similar to the one running Jenkins? I understand that one has a two port card with one ib and one ethernet port. By default both ports will be used for large messages. Not sure why it is affecting small ones. I put together a patch to adjust all the latency and bandwidth numbers for the btls that might help.

**Comment by hjelmn** [comment]
BTW, ~5% is not a blocker. I have seen larger variation due to changes in icache miss rates :-/

**Comment by bosilca** [comment]
Did we lost the exclusivity ?

**Comment by jladd-mlnx** [comment]
Law of least surprise is respected on Master. Same command line that triggered TCP BTL, loads OpenIB on master. The nightly build is from an internal nightly MLNX build. I'll try with master from Git when I get a chance.

**Comment by hjelmn** [comment]
@bosilca The exclusivity check looks ok to me on master. Will check 2.x.

**Comment by hjelmn** [comment]
v2.x looks ok too. We sort the btls be decreasing exclusivity. Then we add the proc to each btl in that order and only add them to btl_send if there is no send btl endpoint for the proc or the exclusivity is equal.

**Comment by bosilca** [comment]
If preventing the TCP BTL from being used improves the latency, then the exclusivity might not work the way we expect.

**Comment by hjelmn** [comment]
Agreed. It should be preventing the tcp btl from running with that proc. Will have to run through the code to see why this could be happening.

**Comment by jladd-mlnx** [comment]
Tests with master head from GitHub. BIG degradation in Yalla. Significant degredation in OpenIB.

OMPI - Master Head f18d6606daedf6
PML - Yalla
```
$mpirun -np 2 --map-by node --bind-to core -mca pml yalla -mca btl openib,self -x MXM_RDMA_PORTS=mlx5_0:1  -mca btl_openib_if_include mlx5_0:1 /hpc/mtr_scrap/users/joshual/GitHub/ompi-release/osu-micro-benchmarks-5.3/install/libexec/osu-micro-benchmarks/mpi/pt2pt/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.3
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       1.38        1384319.88
2                       2.86        1428626.93
4                       5.70        1425752.08
8                      11.43        1428954.19
16                     22.80        1425162.48
32                     45.35        1417328.02
64                     89.76        1402440.95
128                   175.14        1368296.44
256                   343.19        1340577.75
512                   661.68        1292345.50
1024                 1251.70        1222364.08
2048                 2375.11        1159723.45
4096                 3563.16         869911.51
8192                 4996.50         609924.74
16384                5643.96         344479.77
32768                6020.44         183729.28
65536                6242.64          95255.18
131072               6350.21          48448.25
262144               5124.99          19550.29
524288               5706.46          10884.20
1048576              6031.02           5751.63
2097152              6209.33           2960.84
4194304              6303.17           1502.79
```
Another observation - OpenIB message rates are quite erratic from run to run. This is just a single trial that came back particularly degraded 
OMPI - master f18d6606daedf6
PML - OB1 (OpenIB)
```
$mpirun -np 2 --map-by node --bind-to core -mca pml ob1 -mca btl openib,self -x MXM_RDMA_PORTS=mlx5_0:1  -mca btl_openib_if_include mlx5_0:1 /hpc/mtr_scrap/users/joshual/GitHub/ompi-release/osu-micro-benchmarks-5.3/install/libexec/osu-micro-benchmarks/mpi/pt2pt/osu_mbw_mr
# OSU MPI Multiple Bandwidth / Message Rate Test v5.3
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       0.87         867614.62
2                       5.15        2576510.24
4                      10.35        2588480.55
8                      20.60        2574982.19
16                     41.59        2599279.31
32                     81.21        2537720.88
64                    159.46        2491492.93
128                   311.66        2434836.70
256                   597.60        2334391.34
512                  1175.44        2295776.36
1024                 2166.56        2115778.70
2048                 3771.07        1841340.88
4096                 5084.93        1241439.16
8192                 5870.64         716631.08
16384                5835.08         356144.87
32768                6263.04         191132.95
65536                6394.87          97578.03
131072               6461.46          49297.04
262144               6495.77          24779.38
524288               6512.28          12421.18
1048576              6521.56           6219.44
2097152              6525.88           3111.78
4194304              6528.19           1556.44
```

**Comment by jladd-mlnx** [comment]
@hppritcha Please see this Issue for tracking and comments.

**Comment by jladd-mlnx** [comment]
@gpaulsen @nysal Please be aware.

**Comment by thananon** [comment]
I did some performance test and backtracked to before and after we merge #1397 (MPI_THREAD_MULTIPLE support by default). This is way before request refactoring. Here's the result of the same benchmark. 

I did only a couple run, the performance changed a bit every run (especially on small message), but this is enough to get the idea.

<img width="637" alt="screen shot 2016-06-29 at 3 49 44 pm" src="https://cloud.githubusercontent.com/assets/7318806/16471379/26e55bde-3e11-11e6-8996-4d5405c6c77f.png">

**Comment by jladd-mlnx** [comment]
@thananon Thanks!! Very interesting. With which PML was this data collected?

**Comment by thananon** [comment]
@jladd-mlnx This is ob1. Sorry I forgot to mention.

**Comment by jladd-mlnx** [comment]
@thananon could you please bisect and find the guilty commit?

**Comment by thananon** [comment]
I have to do some work in usnic for @jsquyres but I can probably fire up this tracking in the background. It might take me a while.

**Comment by rhc54** [comment]
Perhaps we shouldn't say "guilty", but rather let's find the commit (or commits) where performance started to erode. Might well find that the performance dropped in a set of spurts over some period of time, so tracking performance over some period of commits (e.g., going back a year and taking a snapshot every month) might be more productive.

**Comment by thananon** [comment]
As a side note, since I plotted this earlier, I might as well put them here. We see different effect on different BTL.

![tcp](https://cloud.githubusercontent.com/assets/7318806/16471910/f31c0c90-3e14-11e6-9a5f-3e104bd303f0.png)
![usnic](https://cloud.githubusercontent.com/assets/7318806/16471916/f9b94d06-3e14-11e6-96dd-72cb858bf804.png)
![vader](https://cloud.githubusercontent.com/assets/7318806/16471917/f9b96f5c-3e14-11e6-8385-85da50168b15.png)

**Comment by hjelmn** [comment]
I should point out this is not a 2.0.0 issue. Please open a separate issue for the bml/r2 bug (tcp being used) and assign it to 2.0.1. This issue is a master only issue and needs to be understood but it is not right now important.

**Comment by hjelmn** [comment]
For what it is worth this looks like a similar regression to the ones we already fixed. Something is taking a multi-threaded path when it shouldn't.

**Comment by hjelmn** [comment]
This problem also show the importance of getting performance testing into the nightly MTT. @jjhursey Are there plans to do performance with the next MTT version?

**Comment by hjelmn** [comment]
@thananon Did you collect each BTL's numbers by only enabling that btl (+self)?

**Comment by jjhursey** [comment]
@hjelmn I'll file a feature request for it. I agree that prioritizing that would be a good idea. Might be worth exploring some design options before the Aug F2F meeting, and discussing issues there.
MTT Issue here:
https://github.com/open-mpi/mtt/issues/445

**Comment by hjelmn** [comment]
@jladd-mlnx Not sure why master is not showing the tcp use. It is most definitely being used as an rdma endpoint. I see no performance regression when I use ```-mca btl self,vader``` on my mac. btl/tcp was never intended to be added as an rdma endpoint in any case except when it is used for send/recv. I am opening a PR to correct this behavior. Please test and see if it has any effect on the regression.

**Comment by thananon** [comment]
@hjelmn Yes

**Comment by hjelmn** [comment]
@thananon Thats what I thought. That suggests this isn't just because the tcp btl gets added to the rdma list of each endpoint or else it would affect 2.0 as well. Something else is going on.

**Comment by artpol84** [comment]
@jladd-mlnx @thananon @hjelmn 
I'm working on bisection. Will report later today or tomorrow.

**Comment by artpol84** [comment]
Ok, so in master regression appeared between

commit 97c9732badcc0d3a558fe21856d3c99306fbb8eb
Author: Nathan Hjelm <hjelmn@lanl.gov>
Date:   Thu Oct 22 09:42:52 2015 -0600

and

commit b253d4e887cc9f0c8e73167c53c8b3386b0eb01d
Author: Edgar Gabriel <gabriel@cs.uh.edu>
Date:   Thu Jan 21 08:30:19 2016 -0600

I'm going to bisect now.

Checking release in parallel

**Comment by artpol84** [comment]
Correction: at  97c9732 in master I see 9% Yalla regression (as we currently observe in the release repo). in the current master there is >50% regression.

**Comment by AndreasKempfNEC** [comment]
I would like to relate my experiences with regard to this bug on an InfiniBand EDR system. When compiled with the components of HPC-X version 1.6.392 (hpcx-v1.6.392-gcc-MLNX_OFED_LINUX-3.2-2.0.0.0-redhat7.2), Open MPI 2.0.1rc1 shows extremely low bandwidths on osu_bw and osu_bibw using the yalla pml:

```
***/mpirun -mca pml yalla -x MXM_TLS=rc,shm,self --nooversubscribe --bind-to core --rank-by core --report-bindings --map-by node -np 2 bash -c 'ulimit -s 10240; ***/osu-5.3-ompi2/mpi/pt2pt/osu_bibw' 2>&1 | tee bug.txt
[******n2:22268] MCW rank 0 bound to socket 0[core 0[hwt 0]]: [B/././././././././././././.][./././././././././././././.]
[******n1:23673] MCW rank 1 bound to socket 0[core 0[hwt 0]]: [B/././././././././././././.][./././././././././././././.]
# OSU MPI Bi-Directional Bandwidth Test v5.3
# Size      Bandwidth (MB/s)
1                       4.24
2                       9.11
4                      18.13
8                      37.49
16                     72.31
32                    143.14
64                    268.83
128                   486.04
256                   890.66
512                  1618.42
1024                 2816.44
2048                 4363.90
4096                 6328.37
8192                 6669.60
16384                5909.41
32768                6450.79
65536                6737.78
131072               1024.69
262144               5424.47
524288               5150.51
1048576              4877.21
2097152              2456.51
4194304               630.69
```

When using -mca pml cm, the bandwidths are much better:

```
***/mpirun -mca pml cm -x MXM_TLS=rc,shm,self --nooversubscribe --bind-to core --rank-by core --report-bindings --map-by node -np 2 bash -c 'ulimit -s 10240; ***/osu-5.3-ompi2/mpi/pt2pt/osu_bibw' 2>&1 | tee cm.txt
[******n2:22287] MCW rank 0 bound to socket 0[core 0[hwt 0]]: [B/././././././././././././.][./././././././././././././.]
[******n1:23703] MCW rank 1 bound to socket 0[core 0[hwt 0]]: [B/././././././././././././.][./././././././././././././.]
# OSU MPI Bi-Directional Bandwidth Test v5.3
# Size      Bandwidth (MB/s)
1                       3.58
2                       7.46
4                      14.93
8                      30.62
16                     60.85
32                    119.15
64                    227.54
128                   414.05
256                   774.09
512                  1416.38
1024                 2425.35
2048                 4415.15
4096                 6865.09
8192                 8744.54
16384               13573.98
32768               17952.49
65536               20776.19
131072              22550.33
262144              23573.60
524288              24150.17
1048576             24429.91
2097152             24567.00
4194304             24314.48
```

After looking around for a fix, I stumbled upon this issue. On a whim, I cherry-picked the following three commits:
7d96f12feccde864435e5ff7e81e09fb6c9fd934
bc973cad30b80aac07830d9be9d0a96810e14463
a4ff9bef6d8a5587de32c668b8cdcfd7b623a9f0
and ran the yalla test again:

```
***/mpirun -mca pml yalla -x MXM_TLS=rc,shm,self --nooversubscribe --bind-to core --rank-by core --report-bindings --map-by node -np 2 bash -c 'ulimit -s 10240; ***/osu-5.3-ompi2/mpi/pt2pt/osu_bibw' 2>&1 | tee fix.txt

[r05i17n2:29940] MCW rank 0 bound to socket 0[core 0[hwt 0]]: [B/././././././././././././.][./././././././././././././.]
[r05i14n1:24305] MCW rank 1 bound to socket 0[core 0[hwt 0]]: [B/././././././././././././.][./././././././././././././.]
# OSU MPI Bi-Directional Bandwidth Test v5.3
# Size      Bandwidth (MB/s)
1                       4.21
2                       8.85
4                      17.76
8                      36.34
16                     73.07
32                    146.38
64                    270.31
128                   494.60
256                   910.82
512                  1652.90
1024                 2892.45
2048                 5082.04
4096                 8079.09
8192                 9811.24
16384               14844.96
32768               18508.68
65536               21177.27
131072              22351.17
262144              22744.60
524288              23118.34
1048576             23253.44
2097152             23823.72
4194304             21047.93
```

This is obviously much better (though still not quite on the level of 1.10.3 or cm). Since this bug report was seemingly only concerned with the message rate, I wanted to add my observations that the bandwidth on EDR suffers, as well. I am surprised that the milestone was set to v3.0.0 for this issue.

**Comment by jladd-mlnx** [comment]
@artpol84 I thought these were already PR-ed into 2.0.1. If not, please PR during your day tomorrow.

@jsquyres @hppritcha These three:

7d96f12feccde864435e5ff7e81e09fb6c9fd934
bc973ca
a4ff9be 

need to go into 2.0.1.

**Comment by artpol84** [comment]
I'm confused:

ompi-release/v2.x is free from this bug:
https://github.com/open-mpi/ompi-release/blob/v2.x/ompi/mca/pml/yalla/pml_yalla_request.h#L28

as well as v1.10:
https://github.com/open-mpi/ompi-release/blob/v1.10/ompi/mca/pml/yalla/pml_yalla_request.h#L25

Can you clarify where did you cherry-picked those commits?

**Comment by artpol84** [comment]
I also checked https://www.open-mpi.org/software/ompi/v2.0/downloads/openmpi-2.0.1rc2.tar.bz2 and I don't understand how could you apply mentioned fixes to this codebase.

**Comment by artpol84** [comment]
I'll do the runtime check and will report back

**Comment by AndreasKempfNEC** [comment]
I am sorry for seeding confusion here. I am probably conflating two separate issues which I now realize.

There were some some bandwidth issues with the released tarball of Open MPI 2.0.1rc1 and I found this issue mentioning problems with the message rate on yalla. In the first bug report here, a possible fix on master was mentioned. Looking at the yalla history:
https://github.com/open-mpi/ompi/commits/master/ompi/mca/pml/yalla
I figured that I would test the last 3 commits would be good to test. I updated the files that were touched and the performance was much better. There is a pull request for these commits but it does not mention any bug report:
https://github.com/open-mpi/ompi/pull/1840

Comparing the history of master and release shows me that these three commits do not seem to be in 2.0.X as of yet:
https://github.com/open-mpi/ompi-release/commits/v2.x/ompi/mca/pml/yalla
https://github.com/open-mpi/ompi-release/tree/v2.x/ompi/mca/pml/yalla

Is the pull request for a different bug? Should I comment somewhere else?

**Comment by artpol84** [comment]
As #1840 states the following commit was causing problems:
https://github.com/open-mpi/ompi/commit/041a6a9f53033a12d1cbf5c1af36cb16c7cdcc36

And this commit is not in the 2.0.x. That's why I wonder how you applied patches from my commits you mentioned.

So the question by saying "I updated the files touched by commits" you mean that you replaced them with ones from master? Or did you actually apply patches?

**Comment by artpol84** [comment]
P.S. I'm sure this is the right place to discuss the problem.

**Comment by AndreasKempfNEC** [comment]
Sorry, I just replaced them. I compared the histories and figured they were identical but, clearly, I did not do a good job. I missed a few differences.

**Comment by artpol84** [comment]
Ok, good that we clarified that!
I will double check in my environment and update here.
I'm still surprised because I was testing release and it wasn't affected by this bug. But it's worth to double check.

**Comment by artpol84** [comment]
And thanks for finding and reporting this!

**Comment by jladd-mlnx** [comment]
bot:milestone:v2.0.1

**Comment by jsquyres** [comment]
After reading all the comments from overnight, I'm confused: is there an issue here or not?

**Comment by artpol84** [comment]
@jsquyres need to verify. Will update once done.

**Comment by artpol84** [comment]
I guess that here we face slightly different problem:
![image](https://cloud.githubusercontent.com/assets/7534910/18135199/47516eee-6fcb-11e6-8221-c20b1b485107.png)
Message rate is good for yalla in this measurements - it outperforms cm on small messages.
However we see bandwidth problems when message size grows.
I can reproduce so will do some analysis.

**Comment by jsquyres** [comment]
@artpol84 Is this on master, v2.x, or both?  Was this behavior in v2.0.0?

**Comment by artpol84** [comment]
this is only in 2.x and here is the substantial reason :):
* [master (good perf)](https://github.com/open-mpi/ompi/blob/master/ompi/mca/pml/yalla/pml_yalla.c#L129)
* [v2.x (bad perf)](https://github.com/open-mpi/ompi-release/blob/v2.x/ompi/mca/pml/yalla/pml_yalla.c#L132)

PR: https://github.com/open-mpi/ompi-release/pull/1352

**Comment by artpol84** [comment]
@AndreasKempfNEC thank you once again.

@jladd-mlnx @jsquyres @hppritcha  instead of merging open-mpi/ompi-release#1352 we can pull https://github.com/open-mpi/ompi/pull/1228 and https://github.com/open-mpi/ompi/pull/1840 to the v2.x.

**Comment by artpol84** [comment]
@jsquyres - yes, this behavior is in the v2.0.0

**Comment by jsquyres** [comment]
open-mpi/ompi-release#1352 seems like a good solution for v2.0.1.  The others can wait until v2.0.2.

**Comment by AndreasKempfNEC** [comment]
@artpol84 Thank you for analyzing and fixing the issue so quickly!

To all of the people participating int his discussion: I apologize again for the initial confusion. I took a few too many shortcuts.

So the original report was, indeed, a different problem that was only coincidentally fixed in one of these commits. What an unfortunate typo, though!

Thanks again!

**Comment by jsquyres** [comment]
fixed in open-mpi/ompi-release#1352

**Comment by bosilca** [comment]
The original issue persist on both master and 2.x. 
![figure_1](https://cloud.githubusercontent.com/assets/642701/18183312/c24852e8-7061-11e6-8271-ba4b6661c976.png)

This figures look at the bandwidth reported by NetPIPE and obtained using different versions of Open MPI on a recent Xeon processor. As far as I understand how the process binding works in Open MPI, and assuming there is no bug in recent versions, all the versions should have the same NUMA binding. A general note: while vader does obtain better message rate for small messages, it's bandwidth is significantly lower than SM across **ALL** versions. Second note: something in master and 2.0 affected our performance in a terrible way, we are almost at 1/2 the bandwidth obtained on 1.8 and 1.10.

Before we move forward with this discussion trying to figure out how to fix this, @thananon rolled back master before we started introducing the request completion improvements and the performance issue was already present.

**Comment by artpol84** [comment]
Here is our data for the benchmark that we were using for the message rate experiments

![image](https://cloud.githubusercontent.com/assets/7534910/18280651/ff5958c8-7483-11e6-9b95-7699a7f6d666.png)

**Comment by jsquyres** [comment]
@bosilca Can you post your reproducer scripty-foo that generates those snazzy graphs and whatnot?  That'll save a bunch of time in running those same tests on my platforms over here (so that I can post comparable results here on this issue).

**Comment by thananon** [comment]
Folks,

I finished bisecting the master and I found the commit. It is 664ecc8f84d47f968e95cc26a51f30f3236d26a1. I tested with both NetPIPE and Artem's benchmark. The result is in the graph below. It will be great if someone can confirm this.

![goodbad](https://cloud.githubusercontent.com/assets/7318806/18296379/46044c90-7477-11e6-83e6-862c0f2ac124.png)

@hjelmn

**Comment by artpol84** [comment]
I guess we also reported that some time ago - when we were actively investigating.

**Comment by artpol84** [comment]
IIRC when I was disabling build-in's the performance was much better. And I guess that I was always measuring with them disabled - that's why I've got good results compared to George's ones.

**Comment by artpol84** [comment]
Right, here is my configure line:
```
./configure --with-platform=contrib/platform/mellanox/optimized --prefix=`pwd`/install --enable-mpi-thread-multiple \
            --disable-builtin-atomics
```
@hjelmn was aware

**Comment by jsquyres** [comment]
Well, this certainly does feel like a smoking gun.  Can we narrow it down a little more / back it up with a narrower view of the data?

Is there a small benchmark that we can run (at a bunch of different sites / on a bunch of different platforms) that compares the built-in atomics with the hand-crafted atomics?

**Comment by jsquyres** [comment]
@hjelmn and I just talked about this on the phone.  He helped me understand how OMPI does atomics a bit better.  Basically, at configure time, we choose between 4 different atomics implementations:

1. GCC inline intrinsic atomics (quite similar / related to GCC C++11's atomics).
1. Old-style GCC/ICC `__sync` inline intrinsic atomics.
  * **NOTE:** This should be slower than option 1 because it's a full memory barrier.  Option 1 lets you be more fine-grained / relaxed than a full memory barrier, because we don't need a full memory barrier in many cases.
1. Our own hand-rolled inline atomics.
1. Our own hand-rolled non-inline atomics.

`configure` on master basically checks for them in the above-listed order.  If you `--disable-builtin-atomics`, it skips options 1 and 2.

`configure` on v2.x still prefers option 3; you have to explicitly `--enable-builtin-atomics` to get options 1 or 2.  Meaning: we haven't rolled the change to default to builtin atomics to v2.x yet (see below).

There was a discussion last May where we decided to move to use option 1, partially for maintainability reasons (the hand-rolled atomics are complicated and not many people fully grok them) and partially for performance reasons.  Here's the commit where @bosilca put the change in (first for OS X, and then for everywhere): https://github.com/open-mpi/ompi/commit/3eaefd79 and https://github.com/open-mpi/ompi/commit/277269b6418dfd579b143ba2b83762a2b5b3b4e4.  That commit unfortunately contained a bug, and did not enable option 1 in all cases.  @hjelmn fixed this bug in https://github.com/open-mpi/ompi/commit/664ecc8f84d47f968e95cc26a51f30f3236d26a1.

What we apparently may not have been careful about was measuring that compiler inline intrinsics actually performed better than option 3.  I.e., the performance analysis was not well quantified to the community.  Keep in mind that this is the explicit reason we didn't roll this change-to-use-option-1-by-default to the v2.x branch -- we knew we needed to do this performance analysis first.  That performance analysis unfortunately took a back seat to lots of other issues (other bug fixes, release schedules, summer vacations, ...etc.), but now we're re-engaging and doing the required analysis.

I hope this explains the request for more data to those who are casually watching this issue: we need to measure the performance of options 1 through 3.  If it turns out that option 1 is not uniformly better than option 3, then perhaps we need to revert 664ecc8f84d47f968e95cc26a51f30f3236d26a1, 277269b6418dfd579b143ba2b83762a2b5b3b4e4, and https://github.com/open-mpi/ompi/commit/3eaefd79, and try again in a year or two to see if compiler inline intrinsics have gotten any better.

**Sidenote:** there is an ongoing parallel discussion of dropping option 4 because, at least in some cases, we *know* that the hand-rolled assembly is incorrect.  All compilers that we care about support inline assembly, so options 1 through 3 are sufficient.

**Comment by jladd-mlnx** [comment]
Thanks, Arm!

Do we understand the degradation in performance in going from 1.8.8 to 2.x.
Is this expected? Is it because OpenIB is not thread safe in the 1.8
series?

Josh

**Comment by jsquyres** [comment]
@thananon Isn't that to be expected?  I.e., `sm`/`vader` have, by far, the lowest latency, so they would be most affected by small changes in overall performance.  The same small changes may not be apparent in other BTLs / transports.

**Comment by bosilca** [comment]
@jsquyres there are 2 things in these results. The first graph highlights that today's master when compiled without build-in atomics, see a significant increase in the message rate (with vader). This confirms what @artpol84 said earlier, our change in atomics is responsible for the drop in performance.

Second, we wanted to see if the atomic changes have any impact outside the SM and vader BTLs. @thananon confirmed with the second figure that the impact of built-in atomics on the OpenIB BTL (which is expected to be the next fastest after sm and vader) is minimal. However, there results unfortunately confirm that between 1.8 and master something has been changed in the OpenIB BTL that negatively affect the performance. As @jladd-mlnx pinpointed, the fact that the current BTL is thread-safe when the one in the 1.8 series was not, might explain some performance degradation, but here we are talking almost 20%.

**Comment by thananon** [comment]
@jsquyres Yes, for sm/vader it is a done deal. We found the place where performance degraded. However there's still some performance degradation over IB that is still there and I want to find out why.

@jladd-mlnx I see. That explains a lot. However, this test is single threaded (MPI_Init not MPI_Init_thread) and I'm under the impression that the BTL should not use any lock/atomics if the user request thread single.  Maybe I assumed wrong. 

I also see the latency go up between 1.10.3 and master. Is this also expected? If that's expected then there's nothing to investigate here and I will stop the bisecting. Sorry for confusion.

**Comment by artpol84** [comment]
I discovered problem with yalla inter-node communication.

**Comment by jsquyres** [comment]
Per discussion on the call today: @bosilca thinks that the openib BTL still has a performance degradation issue.

@larrystevenwise @bharatpotnuri Can you look into this?

**Comment by larrystevenwise** [comment]
We can see if this openib issue is seen on iw_cxgb4.  If not, then we should get MLNX help too...

**Comment by larrystevenwise** [comment]
@bharatpotnuri will produce message rate numbers for iw_cxgb4/openib/ob1, 1.10.3 vs 2.0 vs trunk.

**Comment by bharatpotnuri** [comment]
Here are the stats with openmpi-1.10.4 with iwarp :
```
[root@stevo3 ~]# /usr/mpi/gcc/openmpi-1.10.4/bin/mpirun -np 2 --map-by node --display-allocation --allow-run-as-root --hostfile /root/mpd.hosts --mca pml ob1 --prefix /usr/mpi/gcc/openmpi-1.10.4/ --mca btl openib,sm,self /usr/mpi/gcc/openmpi-1.10.4/tests/OSU/osu_mbw_mr

# OSU MPI Multiple Bandwidth / Message Rate Test v3.6
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       2.75        2745602.75
2                       5.50        2747960.49
4                      10.42        2605863.17
8                      15.54        1942926.54
16                     31.22        1951219.51
32                     62.12        1941158.64
64                    122.05        1907032.18
128                   238.83        1865889.21
256                   451.23        1762599.83
512                   806.30        1574803.16
1024                  975.38         952522.70
2048                 1048.58         512000.00
4096                 1112.34         271566.17
8192                 1134.99         138549.13
16384                2210.09         134893.03
32768                2271.37          69316.58
65536                2297.43          35056.01
131072               2315.98          17669.55
262144               2325.16           8869.79
524288               2329.80           4443.73
1048576              2332.13           2224.09
2097152              2333.29           1112.60
4194304              2333.90            556.45
```

Here are the stats with openmpi-2.0.1 with iwarp:
```
[root@stevo3 ~]# /usr/mpi/gcc/openmpi-2.0.1/bin/mpirun -np 2 --map-by node --display-allocation --allow-run-as-root --hostfile /root/mpd.hosts --mca pml ob1 --prefix /usr/mpi/gcc/openmpi-2.0.1/ --mca btl openib,sm,self /usr/mpi/gcc/openmpi-2.0.1/tests/OSU/osu_mbw_mr

# OSU MPI Multiple Bandwidth / Message Rate Test v3.6
# [ pairs: 1 ] [ window size: 64 ]
# Size                  MB/s        Messages/s
1                       2.88        2875242.78
2                       5.79        2894463.34
4                      10.77        2691440.22
8                      15.53        1941329.39
16                     31.36        1960002.48
32                     62.54        1954435.40
64                    122.54        1914713.85
128                   240.18        1876412.45
256                   445.27        1739348.33
512                   806.11        1574428.52
1024                  975.97         953093.22
2048                 1049.15         512278.47
4096                 1112.24         271542.76
8192                 1134.82         138528.35
16384                1515.58          92503.82
32768                1811.36          55278.18
65536                2045.94          31218.56
131072               2177.19          16610.68
262144               2252.87           8594.01
524288               2293.94           4375.35
1048576              2312.94           2205.79
2097152              2323.28           1107.83
4194304              2329.30            555.35

```

**Comment by larrystevenwise** [comment]
Hmm, maybe we need to run on 40GbE T580s.  @bharatpotnuri - do you have access to some T580 nodes?

**Comment by larrystevenwise** [comment]
@bharatpotnuri - you can use r9/r10 tomorrow your time and try the same experiment on the T580s.  I'll email you details. thnx

**Comment by bharatpotnuri** [comment]
@larrystevenwise r9/r10 are gen 2 machines and i got the same numbers as stevo3/4. shall try on stevo1/2 as per our discussion.

**Comment by larrystevenwise** [comment]
@bharatpotnuri - I'll make stevo1/stevo2 ready for you to try on Monday your time.  Thanks.

**Comment by rhc54** [comment]
woohoo!

**Comment by artpol84** [comment]
@thananon can you clarify what 'np' stands for on the diagram?
How many nodes was used? Is this a one-node experiment?
Was there an oversubscription that was under optimiation IIRC.

**Comment by bosilca** [comment]
The information is on the title of the figure. np stands for number of nodes. On each node, the heavily multi-threaded MADNESS application is using all the available cores (in this environment there were 20 cores per node, divided in 19 threads for computations + sends and one for draining the network).

**Comment by artpol84** [comment]
My apologies I haven't noticed info on the diagram

**Comment by artpol84** [comment]
I see that improvement over 1.10 decreases significantly with np increase. Is this because of insufficient amount of computation? Do you know the reason?

**Comment by bosilca** [comment]
Yes, the problem is not computational intensive, but instead communication intensive. Larger scale experiments are under way, however most results will remain private until we publish our work.

**Comment by bharatpotnuri** [comment]
@jsquyres @larrystevenwise  I ran the same tests on stevo1/2 40G interfaces for ompi-1.10.3 and master as well. I see the same stats for both ompi-1.10.3 and master.

**Comment by bosilca** [comment]
Based on the comments here I think it is now safe to assume that the performance degradation for single-threaded applications has been identified and fixed. We can now claim pretty decent performance on [at least] one multi-threaded application and some benchmarks, but this topic is outside of the original topic for this issues, and we should continue the discussion in a more relevant place.

**Comment by rhc54** [comment]
Sorry to re-open this issue, but our latest tests still show significant degradation in rates from 1.10.4 at large message sizes. I cannot provide detailed numbers, but here are the general results. All numbers are relative to 1.10.4. Negative values indicate that the specified version is slower than 1.10.4, while positive values indicate it is faster than OMPI v1.10.4

```
|        msg size     |                   OMPI                  |
|                     |    v2.0.1    |   v2.0.x   |    master   |
-----------------------------------------------------------------
|        < 32         |     -10%     |     -8%    |     +15%    | 
|     32 to 8k        |      -8%     |     -1%    |     -10%    |
|     8k to 128k      |      -1%     |     -1%    |     +20%    |
|        > 128K       |        0     |       0    |      -5%    |
```
Any suggestions?

**Comment by hppritcha** [comment]
what pml was used to obtain these results?

**Comment by rhc54** [comment]
It would be the CM pml

**Comment by bosilca** [comment]
which MTL ?

**Comment by rhc54** [comment]
I believe the numbers are pretty repeatable with either the psm2 or libfabric MTLs. It may just be tuning parameters, of course, or it may be something specific to CM and the integration to the MTL. However, those things didn't change much across the versions (at least, so far as I know) - and perhaps that is the problem, given all the other changes.

**Comment by hppritcha** [comment]
these were probably measured with PSM2 MTL.

**Comment by hppritcha** [comment]
I need more info on how to reproduce this problem.  Was osu_mbw_mr used to generate the message rates for example?  And were any particular PSM2 related env. variables set?

**Comment by matcabral** [comment]
Hi all, 
I got the numbers Ralph is sharing (separate email thread)  by building from ompi's different tags with: 
CFLAGS=-O3 ./configure  --with-libfabric=no   --with-psm2=/usr 
Yes, this is running on PSM2, 2 nodes over a switch , no specific env vars. 

 mpirun -np 2  -host host-1,host-2   ./osu_mbw_mr 
(I did confirm it was runing over PSM2)

**Comment by hppritcha** [comment]
I collected some osu_mbw_mr numbers off of one of the LANL omnipath systems and put them
here https://gist.github.com/hppritcha/8da4436a216c9d96bf10ad9d403da03f

It appears there was a significant performance degradation going from 1.10 release stream to master.  It looks like 2.0.x release stream is the worst, then something was done on master to patch things up a bit, with PRs back to 2.x after the 2.0.x branch was created.

@jsquyres might want to take a look.

**Comment by hppritcha** [comment]
I got data using the GNI provider.  It also shows a significant performance degration for shorter messages.

https://gist.github.com/hppritcha/200bc7a2d4dfde709245d6ffa7b2b971

not as bad as for the PSM2 MTL though.

**Comment by rhc54** [comment]
I don't see the 1.10 data in that gist, but the 2.0.x data is clearly impacted relative to master and v2.x. So it looks like there is something in the code path above the libraries.

@matcabral @hjelmn Can someone take a look and see if something is missing in the CM PML, or related code?

**Comment by hjelmn** [comment]
pml/cm looks up to date to me.

**Comment by rhc54** [comment]
Given that the problem is in master as well, my comment was more to the point that perhaps some change is required in that code path - something that was done to resolve the performance issue in the pml/ob1 path, but didn't get done in the pml/cm path.

**Comment by jsquyres** [comment]
@rhc54 I don't think @hppritcha plans to get the Cray code to work with v1.10.

@rhc54 @matcabral Any progress on this issue?

**Comment by rhc54** [comment]
Still under investigation. I am removing the blocker tag from it as our folks are engaged in some other things right now, and this shouldn't hold up v2.0.2.

@matcabral I spoke with @hjelmn about this at SC, and he suggested looking at the instruction cache for "misses". We have previously seen cases where slight changes to the code path, even when reducing instruction count, would result in performance degradation due to a sudden spike in cache misses. This could be what is happening here.

**Comment by bosilca** [comment]
@thananon is reviewing the CM's code to remove the request lock left over from the transition. The work can be followed on issue #2448.

**Comment by jsquyres** [comment]
How are we doing on this issue on the v2.x branch these days?  Is this still an open issue, or should we close it?

**Comment by thananon** [comment]
I'm not sure if this is fixed. I'm observing performance degradation on my test with builtin atomics. 

This is injection rate in msg/s from multithreaded benchmark.
Compiled with gcc 6.3.0.

|no-bulitin-atomics|default|percentage|
|---|---|---|
|109711|18962|-82%|

**Comment by jsquyres** [comment]
Just to be clear: that was on v2.x, right?

Your table implies that we should switch the default to use the no-builtin-atomics...?

If we're seeing an 82% performance degradation, this feels like a blocker for v2.1.0.  @hppritcha?

**Comment by thananon** [comment]
@jsquyres 
- Test was performed on master.
- Yes, I think it should be no-bultin-atomics by default but that only based on UTK's machine. I don't know about the behavior on others.

**Comment by jsquyres** [comment]
Could you repeat the test on v2.x, please?

**Comment by hppritcha** [comment]
@thananon what BTL are you using when you see the 82 per cent degradation?

**Comment by thananon** [comment]
@jsquyres I will re run the test on 2.x as soon as I can.
@hppritcha I'm using openib btl. However the retest result is inconsistent with the original one. I will redo the whole test and report back as soon as possible.

**Comment by jsquyres** [comment]
Per the OMPI webex this morning, we're deferring this to v2.1.1.

**Comment by hppritcha** [comment]
Has this actually been fixed in 2.1.1?  Otherwise I'd like to move this to future milestone.

**Comment by matcabral** [comment]
the problems I found were addressed in #3748
thanks

**Comment by hppritcha** [comment]
per last comment from @matcabral closing this issue
